README.md
# GalleryImageViewer

An easy way to display images in a full-screen dialog, including pinch & zoom.

[![Pub](https://img.shields.io/pub/v/gallery_image_viewer.svg)](https://pub.dartlang.org/packages/gallery_image_viewer)
[![Tests](https://github.com/SEMAJdev/gallery_image_viewer/actions/workflows/dart.yml/badge.svg)](https://github.com/SEMAJdev/gallery_image_viewer/actions/workflows/dart.yml)

![Easy Image Viewer Demo](https://github.com/SEMAJdev/gallery_image_viewer/blob/main/demo_images/demo_gallery_image_viewer1.gif?raw=true "Easy Image Viewer Demo")

## Features

* Show a single image or a swipeable list of images
* Use pinch & zoom to zoom in and out of images
* Optionally allow "swipe down to dismiss" by passing in `swipeDismissible: true`
* No dependencies besides Flutter
* Callbacks for `onPageChanged` and `onViewerDismissed`

## Usage

Show a Gallery image:
```dart
final List<ImageProvider> _imageProviders = [
    Image.network("https://picsum.photos/id/237/200/300").image,
    Image.network("https://picsum.photos/seed/picsum/200/300").image,
    Image.network("https://picsum.photos/200/300?grayscale").image,
    Image.network("https://picsum.photos/200/300").image,
    Image.network("https://picsum.photos/200/300?grayscale").image
  ];
GalleryImageView(
  listImage: _imageProviders,
  width: 200,
  height: 200,
  imageDecoration: BoxDecoration(border: Border.all(color: Colors.white)),
),
```


Show a single image:

```dart
final imageProvider = Image.network("https://picsum.photos/id/1001/5616/3744").image;
showImageViewer(context, imageProvider, onViewerDismissed: () {
  print("dismissed");
});
```

Show a bunch of images:

```dart
MultiImageProvider multiImageProvider = MultiImageProvider([
  Image.network("https://picsum.photos/id/1001/5616/3744").image,
  Image.network("https://picsum.photos/id/1003/1181/1772").image,
  Image.network("https://picsum.photos/id/1004/5616/3744").image,
  Image.network("https://picsum.photos/id/1005/5760/3840").image
]);

showImageViewerPager(context, multiImageProvider, onPageChanged: (page) {
  print("page changed to $page");
}, onViewerDismissed: (page) {
  print("dismissed while on page $page");
});
```

Usually you'll want to implement your own `EasyImageProvider`. Suppose you have
a list of `Product`s, each of which has an `imagePath` property with the path
to a local image file. You could create an `EasyImageProvider` that takes a list
of `Product`s like this:

```dart
class ProductsImageProvider extends EasyImageProvider {

  final List<Product> products;
  final int initialIndex;

  ProductsImageProvider({ required this.products, this.initialIndex = 0 });

  @override
  ImageProvider<Object> imageBuilder(BuildContext context, int index) {
    String? localImagePath = products[index].imagePath;
    File? imageFile;

    if (localImagePath != null) {
      imageFile = File(localImagePath);
    }

    ImageProvider imageProvider = imageFile != null ? FileImage(imageFile) : AssetImage("assets/images/product_placeholder.jpg") as ImageProvider;

    return imageProvider;
  }

  @override
  int get imageCount => products.length;
}
```

You could then use it like this:

```dart
ProductsImageProvider productsImageProvider = ProductsImageProvider(products: products);

showImageViewerPager(context, productsImageProvider, onPageChanged: (page) {
  print("page changed to $page");
}, onViewerDismissed: (page) {
  print("dismissed while on page $page");
});
```

## Credits
GalleryImageViewer is custom from EasyImageViewer.
EasyImageViewer is a project by [TSG](https://thesmythgroup.com/), a full-service digital agency taking software from concept to launch.
Our powerhouse team of designers and engineers build iOS, Android, and web apps across many industries.


CHANGELOG.md
# EasyImageViewer Changelog

## 1.1.0

- Update demo image path in README.

## 1.1.1

- 20/30 points: code has no errors, warnings, lints, or formatting issues
  lib/src/gallery_image_view.dart doesn't match the Dart formatter.
  To format your files run: flutter format .


## 1.2.0
- Update README.

test/easy_image_view_test.dart
import 'package:gallery_image_viewer/src/easy_image_view.dart';

import 'support/test_helper.dart';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('EasyImageView', () {
    testWidgets('should have an image and a scale',
        (WidgetTester tester) async {
      ImageProvider? imageProvider;

      await tester.runAsync(() async {
        imageProvider = await createColorImage(Colors.green);
      });

      Widget testWidget = MediaQuery(
          data: const MediaQueryData(size: Size(600, 800)),
          child: EasyImageView(
              imageProvider: imageProvider!, minScale: 0.5, maxScale: 6.0));

      await tester.pumpWidget(testWidget);

      // Create the Finders.
      final sizedBoxFinder = find.byKey(const Key('easy_image_sized_box'));
      final interactiveViewFinder =
          find.byKey(const Key('easy_image_interactive_viewer'));
      final imageFinder = find.image(imageProvider!);

      // Check existence
      expect(sizedBoxFinder, findsOneWidget);
      expect(interactiveViewFinder, findsOneWidget);
      expect(imageFinder, findsOneWidget);

      // Check properties
      SizedBox sizedBox = tester.firstWidget(sizedBoxFinder);
      expect(sizedBox.width, double.infinity);
      expect(sizedBox.height, double.infinity);
      InteractiveViewer interactiveViewer =
          tester.firstWidget(interactiveViewFinder);
      expect(interactiveViewer.minScale, 0.5);
      expect(interactiveViewer.maxScale, 6.0);
    });

    testWidgets('should invoke the onScaleChanged callback',
        (WidgetTester tester) async {
      ImageProvider? imageProvider;

      double lastScale = 1.0;

      await tester.runAsync(() async {
        imageProvider = await createColorImage(Colors.green);
      });

      Widget testWidget = MediaQuery(
          data: const MediaQueryData(size: Size(600, 800)),
          child: EasyImageView(
              imageProvider: imageProvider!,
              minScale: 0.5,
              maxScale: 6.0,
              onScaleChanged: (scale) {
                lastScale = scale;
              }));

      await tester.pumpWidget(testWidget);

      // Create the finder
      final interactiveViewFinder = find.byType(InteractiveViewer);

      // Get the center
      final center = tester.getCenter(interactiveViewFinder);

      // Zoom in:
      final Offset scaleStart1 = center;
      final Offset scaleStart2 = Offset(center.dx + 10.0, center.dy);
      final Offset scaleEnd1 = Offset(center.dx - 10.0, center.dy);
      final Offset scaleEnd2 = Offset(center.dx + 10.0, center.dy);
      final TestGesture gesture = await tester.createGesture();
      final TestGesture gesture2 = await tester.createGesture();
      await gesture.down(scaleStart1);
      await gesture2.down(scaleStart2);
      await tester.pump();
      await gesture.moveTo(scaleEnd1);
      await gesture2.moveTo(scaleEnd2);
      await tester.pump();
      await gesture.up();
      await gesture2.up();
      await tester.pumpAndSettle();

      expect(lastScale, 2.0);
    });
  });
}


test/easy_image_viewer_test.dart
import 'package:gallery_image_viewer/gallery_image_viewer.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'support/test_helper.dart';

void main() {
  group('showImageViewer', () {
    testWidgets(
        'should have a PageView of images and invoke dismissal callback',
        (WidgetTester tester) async {
      ImageProvider? redImageProvider;
      BuildContext context = await createTestBuildContext(tester);
      bool dismissed = false;

      await tester.runAsync(() async {
        redImageProvider = await createColorImage(Colors.red);
      });

      final dialogFuture =
          showImageViewer(context, redImageProvider!, onViewerDismissed: () {
        dismissed = true;
      });
      await tester.pumpAndSettle();

      // Create the Finders.
      final pageViewFinder =
          find.byWidgetPredicate((widget) => widget is PageView);
      final closeButtonFinder = find.byIcon(Icons.close);

      // Check existence
      expect(pageViewFinder, findsOneWidget);
      expect(closeButtonFinder, findsOneWidget);

      // Dismiss the dialog
      await tester.tap(closeButtonFinder);

      await dialogFuture;

      expect(dismissed, true);
    });

    testWidgets('should respect the backgroundColor and closeButtonColor',
        (WidgetTester tester) async {
      late ImageProvider imageProvider;
      final context = await createTestBuildContext(tester);

      await tester.runAsync(() async {
        imageProvider = await createColorImage(Colors.amber);
      });

      showImageViewer(context, imageProvider,
          backgroundColor: Colors.red, closeButtonColor: Colors.green);
      await tester.pumpAndSettle();

      // Check default closeButtonColor
      IconButton closeButton =
          tester.firstWidget(find.widgetWithIcon(IconButton, Icons.close));
      expect(closeButton.color, Colors.green);

      // Check default dialog backgroundColor
      Dialog dialog = tester
          .firstWidget(find.byWidgetPredicate((widget) => widget is Dialog));
      expect(dialog.backgroundColor, Colors.red);
    });
  });

  group('showImageViewerPager', () {
    testWidgets('should have a PageView of images and invoke callbacks',
        (WidgetTester tester) async {
      List<ImageProvider> imageProviders = List.empty(growable: true);
      final context = await createTestBuildContext(tester);
      bool dismissed = false;
      int currentPage = -1;
      int pageOnDismissal = -1;

      await tester.runAsync(() async {
        const colors = [
          Colors.amber,
          Colors.red,
          Colors.green,
          Colors.blue,
          Colors.teal
        ];
        imageProviders =
            await Future.wait(colors.map((color) => createColorImage(color)));
      });

      final multiImageProvider = MultiImageProvider(imageProviders);

      final dialogFuture = showImageViewerPager(context, multiImageProvider,
          onPageChanged: (page) {
        currentPage = page;
      }, onViewerDismissed: (page) {
        dismissed = true;
        pageOnDismissal = page;
      });
      await tester.pumpAndSettle();

      // Create the Finders.
      final pageViewFinder = find.byKey(GlobalObjectKey(multiImageProvider));
      final closeButtonFinder = find.byIcon(Icons.close);

      // Check existence
      expect(pageViewFinder, findsOneWidget);
      expect(closeButtonFinder, findsOneWidget);

      // Swipe to second image
      await tester.drag(pageViewFinder, const Offset(-501.0, 0.0));
      await tester.pumpAndSettle();
      expect(currentPage, 1);

      // Swipe to third image
      await tester.drag(pageViewFinder, const Offset(-501.0, 0.0));
      await tester.pumpAndSettle();
      expect(currentPage, 2);

      // Dismiss the dialog
      await tester.tap(closeButtonFinder);

      await dialogFuture;

      expect(dismissed, true);
      expect(pageOnDismissal, 2);
    });

    testWidgets('should invoke callbacks when dismissed with a swipe',
        (WidgetTester tester) async {
      List<ImageProvider> imageProviders = List.empty(growable: true);
      final context = await createTestBuildContext(tester);
      bool dismissed = false;
      int pageOnDismissal = -1;

      await tester.runAsync(() async {
        const colors = [
          Colors.amber,
          Colors.red,
          Colors.green,
          Colors.blue,
          Colors.teal
        ];
        imageProviders =
            await Future.wait(colors.map((color) => createColorImage(color)));
      });

      final multiImageProvider = MultiImageProvider(imageProviders);

      final dialogFuture = showImageViewerPager(context, multiImageProvider,
          swipeDismissible: true, onViewerDismissed: (page) {
        dismissed = true;
        pageOnDismissal = page;
      });
      await tester.pumpAndSettle();

      // Create the Finders.
      final dismissibleFinder =
          find.byWidgetPredicate((widget) => widget is Dismissible);

      // Check existence
      expect(dismissibleFinder, findsOneWidget);

      // Swipe to dismiss
      await tester.drag(dismissibleFinder, const Offset(0, 501.0));
      await tester.pumpAndSettle();

      await dialogFuture;

      expect(dismissed, true);
      expect(pageOnDismissal, 0);
    });

    testWidgets('should respect the initialIndex', (WidgetTester tester) async {
      List<ImageProvider> imageProviders = List.empty(growable: true);
      final context = await createTestBuildContext(tester);
      bool dismissed = false;
      int pageOnDismissal = -1;

      await tester.runAsync(() async {
        const colors = [
          Colors.amber,
          Colors.red,
          Colors.green,
          Colors.blue,
          Colors.teal
        ];
        imageProviders =
            await Future.wait(colors.map((color) => createColorImage(color)));
      });

      final multiImageProvider =
          MultiImageProvider(imageProviders, initialIndex: 2);

      final dialogFuture = showImageViewerPager(context, multiImageProvider,
          onViewerDismissed: (page) {
        dismissed = true;
        pageOnDismissal = page;
      });
      await tester.pumpAndSettle();

      // Create the Finders.
      final closeButtonFinder = find.byIcon(Icons.close);

      // Check default closeButtonColor
      IconButton closeButton =
          tester.firstWidget(find.widgetWithIcon(IconButton, Icons.close));
      expect(closeButton.color, Colors.white);

      // Check default dialog backgroundColor
      Dialog dialog = tester
          .firstWidget(find.byWidgetPredicate((widget) => widget is Dialog));
      expect(dialog.backgroundColor, Colors.black);

      // Dismiss the dialog
      await tester.tap(closeButtonFinder);

      await dialogFuture;

      expect(dismissed, true);
      expect(pageOnDismissal, 2);
    });

    testWidgets('should respect the backgroundColor and closeButtonColor',
        (WidgetTester tester) async {
      List<ImageProvider> imageProviders = List.empty(growable: true);
      final context = await createTestBuildContext(tester);

      await tester.runAsync(() async {
        const colors = [Colors.amber];
        imageProviders =
            await Future.wait(colors.map((color) => createColorImage(color)));
      });

      final multiImageProvider = MultiImageProvider(imageProviders);

      showImageViewerPager(context, multiImageProvider,
          backgroundColor: Colors.red, closeButtonColor: Colors.green);
      await tester.pumpAndSettle();

      // Check default closeButtonColor
      IconButton closeButton =
          tester.firstWidget(find.widgetWithIcon(IconButton, Icons.close));
      expect(closeButton.color, Colors.green);

      // Check default dialog backgroundColor
      Dialog dialog = tester
          .firstWidget(find.byWidgetPredicate((widget) => widget is Dialog));
      expect(dialog.backgroundColor, Colors.red);
    });
  });
}


test/multi_image_provider_test.dart
import 'package:gallery_image_viewer/src/multi_image_provider.dart';

import 'support/test_helper.dart';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('MultiImageProvider', () {
    test('should require a valid initialIndex', () async {
      final imageProviders = [
        await createColorImage(Colors.red),
        await createColorImage(Colors.green)
      ];

      expect(() => MultiImageProvider(imageProviders, initialIndex: -1),
          throwsArgumentError);
      expect(() => MultiImageProvider(imageProviders, initialIndex: 2),
          throwsArgumentError);

      final provider0 = MultiImageProvider(imageProviders, initialIndex: 0);
      final provider1 = MultiImageProvider(imageProviders, initialIndex: 1);

      expect(provider0.initialIndex, 0);
      expect(provider1.initialIndex, 1);
    });

    testWidgets('should return the correct image per index',
        (WidgetTester tester) async {
      ImageProvider? redImageProvider;
      ImageProvider? greenImageProvider;
      BuildContext context = await createTestBuildContext(tester);

      await tester.runAsync(() async {
        redImageProvider = await createColorImage(Colors.red);
        greenImageProvider = await createColorImage(Colors.green);
      });

      final imageProviders = [redImageProvider!, greenImageProvider!];
      final provider = MultiImageProvider(imageProviders, initialIndex: 0);

      expect(provider.imageCount, 2);

      expect(provider.imageBuilder(context, 0), redImageProvider);
      expect(provider.imageBuilder(context, 1), greenImageProvider);
      expect(() => provider.imageBuilder(context, -1), throwsArgumentError);
      expect(() => provider.imageBuilder(context, 2), throwsArgumentError);
    });

    test('should require a non-empty list of ImageProviders', () async {
      expect(() => MultiImageProvider([]), throwsArgumentError);
    });
  });
}


lib/easy_image_viewer.dart
/// A library to easily display images in a full-screen dialog.
/// It supports pinch & zoom, and paging through multiple images.
library easy_image_viewer;

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import 'src/easy_image_provider.dart';
import 'src/easy_image_viewer_dismissible_dialog.dart';
import 'src/single_image_provider.dart';

export 'src/easy_image_provider.dart' show EasyImageProvider;
export 'src/single_image_provider.dart' show SingleImageProvider;
export 'src/multi_image_provider.dart' show MultiImageProvider;

export 'src/easy_image_view.dart' show EasyImageView;
export 'src/easy_image_view_pager.dart' show EasyImageViewPager;
export 'src/gallery_image_view.dart' show GalleryImageView;

// Defined here so we don't repeat ourselves
const _defaultBackgroundColor = Colors.black;
const _defaultCloseButtonColor = Colors.white;
const _defaultCloseButtonTooltip = 'Close';

/// Shows the given [imageProvider] in a full-screen [Dialog].
/// Setting [immersive] to false will prevent the top and bottom bars from being hidden.
/// The optional [onViewerDismissed] callback function is called when the dialog is closed.
/// The optional [useSafeArea] boolean defaults to false and is passed to [showDialog].
/// The optional [swipeDismissible] boolean defaults to false allows swipe-down-to-dismiss.
/// The [backgroundColor] defaults to black, but can be set to any other color.
/// The [closeButtonTooltip] text is displayed when the user long-presses on the
/// close button and is used for accessibility.
/// The [closeButtonColor] defaults to white, but can be set to any other color.
Future<Dialog?> showImageViewer(
    BuildContext context, ImageProvider imageProvider,
    {bool immersive = true,
    void Function()? onViewerDismissed,
    bool useSafeArea = false,
    bool swipeDismissible = false,
    Color backgroundColor = _defaultBackgroundColor,
    String closeButtonTooltip = _defaultCloseButtonTooltip,
    Color closeButtonColor = _defaultCloseButtonColor}) {
  return showImageViewerPager(context, SingleImageProvider(imageProvider),
      immersive: immersive,
      onViewerDismissed:
          onViewerDismissed != null ? (_) => onViewerDismissed() : null,
      useSafeArea: useSafeArea,
      swipeDismissible: swipeDismissible,
      backgroundColor: backgroundColor,
      closeButtonTooltip: closeButtonTooltip,
      closeButtonColor: closeButtonColor);
}

/// Shows the images provided by the [imageProvider] in a full-screen PageView [Dialog].
/// Setting [immersive] to false will prevent the top and bottom bars from being hidden.
/// The optional [onPageChanged] callback function is called with the index of
/// the image when the user has swiped to another image.
/// The optional [onViewerDismissed] callback function is called with the index of
/// the image that is displayed when the dialog is closed.
/// The optional [useSafeArea] boolean defaults to false and is passed to [showDialog].
/// The optional [swipeDismissible] boolean defaults to false allows swipe-down-to-dismiss.
/// The [backgroundColor] defaults to black, but can be set to any other color.
/// The [closeButtonTooltip] text is displayed when the user long-presses on the
/// close button and is used for accessibility.
/// The [closeButtonColor] defaults to white, but can be set to any other color.
Future<Dialog?> showImageViewerPager(
    BuildContext context, EasyImageProvider imageProvider,
    {bool immersive = true,
    void Function(int)? onPageChanged,
    void Function(int)? onViewerDismissed,
    bool useSafeArea = false,
    bool swipeDismissible = false,
    Color backgroundColor = _defaultBackgroundColor,
    String closeButtonTooltip = _defaultCloseButtonTooltip,
    Color closeButtonColor = _defaultCloseButtonColor}) {
  if (immersive) {
    // Hide top and bottom bars
    SystemChrome.setEnabledSystemUIMode(SystemUiMode.immersive);
  }

  return showDialog<Dialog>(
      context: context,
      useSafeArea: useSafeArea,
      builder: (context) {
        return EasyImageViewerDismissibleDialog(imageProvider,
            immersive: immersive,
            onPageChanged: onPageChanged,
            onViewerDismissed: onViewerDismissed,
            useSafeArea: useSafeArea,
            swipeDismissible: swipeDismissible,
            backgroundColor: backgroundColor,
            closeButtonColor: closeButtonColor,
            closeButtonTooltip: closeButtonTooltip);
      });
}


lib/gallery_image_viewer.dart
/// A library to easily display images in a full-screen dialog.
/// It supports pinch & zoom, and paging through multiple images.
library gallery_image_viewer;

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import 'src/easy_image_provider.dart';
import 'src/easy_image_viewer_dismissible_dialog.dart';
import 'src/single_image_provider.dart';

export 'src/easy_image_provider.dart' show EasyImageProvider;
export 'src/single_image_provider.dart' show SingleImageProvider;
export 'src/multi_image_provider.dart' show MultiImageProvider;

export 'src/easy_image_view.dart' show EasyImageView;
export 'src/easy_image_view_pager.dart' show EasyImageViewPager;
export 'src/gallery_image_view.dart' show GalleryImageView;

// Defined here so we don't repeat ourselves
const _defaultBackgroundColor = Colors.black;
const _defaultCloseButtonColor = Colors.white;
const _defaultCloseButtonTooltip = 'Close';

/// Shows the given [imageProvider] in a full-screen [Dialog].
/// Setting [immersive] to false will prevent the top and bottom bars from being hidden.
/// The optional [onViewerDismissed] callback function is called when the dialog is closed.
/// The optional [useSafeArea] boolean defaults to false and is passed to [showDialog].
/// The optional [swipeDismissible] boolean defaults to false allows swipe-down-to-dismiss.
/// The [backgroundColor] defaults to black, but can be set to any other color.
/// The [closeButtonTooltip] text is displayed when the user long-presses on the
/// close button and is used for accessibility.
/// The [closeButtonColor] defaults to white, but can be set to any other color.
Future<Dialog?> showImageViewer(
    BuildContext context, ImageProvider imageProvider,
    {bool immersive = true,
    void Function()? onViewerDismissed,
    bool useSafeArea = false,
    bool swipeDismissible = false,
    Color backgroundColor = _defaultBackgroundColor,
    String closeButtonTooltip = _defaultCloseButtonTooltip,
    Color closeButtonColor = _defaultCloseButtonColor}) {
  return showImageViewerPager(context, SingleImageProvider(imageProvider),
      immersive: immersive,
      onViewerDismissed:
          onViewerDismissed != null ? (_) => onViewerDismissed() : null,
      useSafeArea: useSafeArea,
      swipeDismissible: swipeDismissible,
      backgroundColor: backgroundColor,
      closeButtonTooltip: closeButtonTooltip,
      closeButtonColor: closeButtonColor);
}

/// Shows the images provided by the [imageProvider] in a full-screen PageView [Dialog].
/// Setting [immersive] to false will prevent the top and bottom bars from being hidden.
/// The optional [onPageChanged] callback function is called with the index of
/// the image when the user has swiped to another image.
/// The optional [onViewerDismissed] callback function is called with the index of
/// the image that is displayed when the dialog is closed.
/// The optional [useSafeArea] boolean defaults to false and is passed to [showDialog].
/// The optional [swipeDismissible] boolean defaults to false allows swipe-down-to-dismiss.
/// The [backgroundColor] defaults to black, but can be set to any other color.
/// The [closeButtonTooltip] text is displayed when the user long-presses on the
/// close button and is used for accessibility.
/// The [closeButtonColor] defaults to white, but can be set to any other color.
Future<Dialog?> showImageViewerPager(
    BuildContext context, EasyImageProvider imageProvider,
    {bool immersive = true,
    void Function(int)? onPageChanged,
    void Function(int)? onViewerDismissed,
    bool useSafeArea = false,
    bool swipeDismissible = false,
    Color backgroundColor = _defaultBackgroundColor,
    String closeButtonTooltip = _defaultCloseButtonTooltip,
    Color closeButtonColor = _defaultCloseButtonColor}) {
  if (immersive) {
    // Hide top and bottom bars
    SystemChrome.setEnabledSystemUIMode(SystemUiMode.immersive);
  }

  return showDialog<Dialog>(
      context: context,
      useSafeArea: useSafeArea,
      builder: (context) {
        return EasyImageViewerDismissibleDialog(imageProvider,
            immersive: immersive,
            onPageChanged: onPageChanged,
            onViewerDismissed: onViewerDismissed,
            useSafeArea: useSafeArea,
            swipeDismissible: swipeDismissible,
            backgroundColor: backgroundColor,
            closeButtonColor: closeButtonColor,
            closeButtonTooltip: closeButtonTooltip);
      });
}


example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.CreateAndShow(L"gallery_image_viewer_example", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


example/lib/main.dart
import 'package:gallery_image_viewer/gallery_image_viewer.dart';
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'GalleryImageViewer Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'GalleryImageViewer Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  final String title;

  const MyHomePage({
    Key? key,
    required this.title,
  }) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final List<ImageProvider> _imageProviders = [
    Image.network("https://picsum.photos/id/237/200/300").image,
    Image.network("https://picsum.photos/seed/picsum/200/300").image,
    Image.network("https://picsum.photos/200/300?grayscale").image,
    Image.network("https://picsum.photos/200/300").image,
    Image.network("https://picsum.photos/200/300?grayscale").image
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SingleChildScrollView(
        child: Center(
            child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            ElevatedButton(
                child: const Text("Show Single Image"),
                onPressed: () {
                  showImageViewer(
                      context,
                      Image.network("https://picsum.photos/id/237/200/300")
                          .image,
                      swipeDismissible: true);
                }),
            ElevatedButton(
                child: const Text("Show Multiple Images (Simple)"),
                onPressed: () {
                  MultiImageProvider multiImageProvider =
                      MultiImageProvider(_imageProviders);
                  showImageViewerPager(context, multiImageProvider,
                      swipeDismissible: true);
                }),
            const Text("<<<< Gallery type 1 >>>>"),
            GalleryImageView(
              listImage: _imageProviders,
              width: 200,
              height: 200,
              imageDecoration:
                  BoxDecoration(border: Border.all(color: Colors.white)),
              galleryType: 1,
            ),
            const SizedBox(
              height: 10,
            ),
            const Text("<<<< Gallery type 2 >>>>"),
            GalleryImageView(
              listImage: _imageProviders,
              width: 200,
              height: 200,
              imageDecoration:
                  BoxDecoration(border: Border.all(color: Colors.white)),
              galleryType: 2,
            ),
            const SizedBox(
              height: 10,
            ),
            const Text("<<<< Gallery type 3 >>>>"),
            GalleryImageView(
              listImage: _imageProviders,
              width: 200,
              height: 200,
              imageDecoration:
                  BoxDecoration(border: Border.all(color: Colors.white)),
              galleryType: 3,
            )
          ],
        )),
      ),
    );
  }
}

class CustomImageProvider extends EasyImageProvider {
  @override
  final int initialIndex;
  final List<String> imageUrls;

  CustomImageProvider({required this.imageUrls, this.initialIndex = 0})
      : super();

  @override
  ImageProvider<Object> imageBuilder(BuildContext context, int index) {
    return NetworkImage(imageUrls[index]);
  }

  @override
  int get imageCount => imageUrls.length;
}
